api/godog/json/matcher.go

```diff
@@ -55,270 +30,76 @@ func Match(json []byte, path string, pattern string) error {
 	}
 
-	return t, nil
-}
-
-func parseTerm(lex *scanner.Scanner) (term, error) {
-	switch lex.Scan() {
-	case '{':
-		return parseObject(lex)
-	case '[':
-		return parseArray(lex)
-	case '-':
-		return parseNegativeNumber(lex)
-	case scanner.Ident:
-		return encodeIdent(lex)
-	case scanner.String:
-		return encodeString(lex)
-	case scanner.Int, scanner.Float:
-		return encodeNumber(lex)
-	case scanner.EOF:
-		return null, errors.Errorf("Parse error: Unexpected EOF")
-	default:
-		return null, errors.Errorf("Parse error: Unexpected token: '%c'", lex.Peek())
-	}
+	return nil
 }
 
-func encodeIdent(lex *scanner.Scanner) (term, error) {
-	text := lex.TokenText()
-
-	switch text {
-	case "null":
-		return null, nil
-	case "true", "false":
-		return term{jsonType: jsonBool, value: text}, nil
-	}
-
-	return null, errors.Errorf("Parse error: Unexpected token: %s", text)
-}
+func Read(json []byte, path string) (result string, err error) {
+	path = revertLegacySyntax(path)
 
-func encodeString(lex *scanner.Scanner) (term, error) {
-	str, err := strconv.Unquote(lex.TokenText())
+	res := gjson.Get(string(json), path)
 
-	if err != nil {
-		return null, err
+	if !res.Exists() || res.IsObject() || res.IsArray() {
+		return "", errors.Errorf("Match error: Expected a scalar got '%T' JSON: %s", res.String(), string(json))
 	}
 
-	return term{jsonType: jsonString, value: str}, nil
+	return res.String(), nil
 }
 
-func encodeNumber(lex *scanner.Scanner) (term, error) {
-	text := lex.TokenText()
-	return term{jsonType: jsonNumber, value: text}, nil
-}
-
-func parseObject(lex *scanner.Scanner) (term, error) {
-	obj := make(map[string]term)
-
-	for tok := lex.Scan(); tok != scanner.EOF && tok != '}'; tok = lex.Scan() {
-		if tok != scanner.String {
-			return null, errors.Errorf("Parse error: Expected a key got: %s", lex.TokenText())
-		}
-
-		key, err := strconv.Unquote(lex.TokenText())
-		if err != nil {
-			return null, err
-		}
+func ReadStringArr(json []byte, path string) (result []string, err error) {
+	path = revertLegacySyntax(path)
 
-		tok = lex.Scan()
-		if tok != ':' {
-			return null, errors.Errorf("Parse error: Expected separator: %s", lex.TokenText())
-		}
+	res := gjson.Get(string(json), path)
 
-		val, err := parseTerm(lex)
-		if err != nil {
-			return null, err
-		}
 
-		if lex.Peek() == ',' {
-			lex.Scan()
-		}
 
-		obj[key] = val
+	if !res.IsArray() {
+		return []string{}, errors.Errorf("Match error: Expected an array got: %s", res.String())
 	}
 
-	return term{jsonType: jsonObject, value: obj}, nil
-}
-
-func parseArray(lex *scanner.Scanner) (term, error) {
-	arr := make([]term, 0)
-
-	for lex.Peek() != ']' {
-		val, err := parseTerm(lex)
-		if err != nil {
-			return null, err
-		}
-
-		arr = append(arr, val)
-
-		if lex.Peek() == ',' {
-			lex.Scan()
+	for _, v := range res.Array() {
+		stringValue := v.String()
+		if !(v.IsObject() || !v.IsArray()) {
```

Comment:
I don't understand this; if something is not an object but an array, we throw an error.

If it's neither an object nor an array then I guess it's a scalar so then I understand it's OK. But if v is an object it's OK even if it's also an array. Should non-scalar objects be accepted at all? Can something can be both an object and an array?

If the logic is correct, maybe a small comment could be useful? Could it maybe be made a little bit clearer by doing  this instead:
```suggestion
		if !v.IsObject() && v.IsArray()) {
```

Reply:
Yes, this was wrong